File: python/tests/test_errors.py
"""Tests for error types."""

from payskill.errors import (
    PayError,
    PayInsufficientFundsError,
    PayNetworkError,
    PayServerError,
    PayValidationError,
)


class TestErrorHierarchy:
    def test_all_inherit_from_pay_error(self) -> None:
        assert issubclass(PayValidationError, PayError)
        assert issubclass(PayNetworkError, PayError)
        assert issubclass(PayServerError, PayError)
        assert issubclass(PayInsufficientFundsError, PayError)

    def test_validation_error_fields(self) -> None:
        err = PayValidationError("bad input", field="amount")
        assert err.code == "validation_error"
        assert err.field == "amount"
        assert "bad input" in str(err)

    def test_server_error_status(self) -> None:
        err = PayServerError("not found", status_code=404)
        assert err.status_code == 404
        assert err.code == "server_error"

    def test_network_error(self) -> None:
        err = PayNetworkError("connection refused")
        assert err.code == "network_error"

    def test_insufficient_funds_hint(self) -> None:
        err = PayInsufficientFundsError("not enough", balance=5.0, required=10.0)
        assert err.code == "insufficient_funds"
        assert err.balance == 5.0
        assert err.required == 10.0
        assert "create_fund_link" in str(err)

    def test_base_error_code(self) -> None:
        err = PayError("generic")
        assert err.code == "pay_error"

File: typescript/tests/test_errors.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  PayError,
  PayValidationError,
  PayNetworkError,
  PayServerError,
  PayInsufficientFundsError,
} from "../src/errors.js";

describe("error hierarchy", () => {
  it("all errors extend PayError", () => {
    assert.ok(new PayValidationError("x") instanceof PayError);
    assert.ok(new PayNetworkError("x") instanceof PayError);
    assert.ok(new PayServerError("x", 400) instanceof PayError);
    assert.ok(new PayInsufficientFundsError("x") instanceof PayError);
  });

  it("PayValidationError has field and code", () => {
    const err = new PayValidationError("bad input", "amount");
    assert.equal(err.code, "validation_error");
    assert.equal(err.field, "amount");
    assert.equal(err.message, "bad input");
  });

  it("PayServerError has statusCode", () => {
    const err = new PayServerError("not found", 404);
    assert.equal(err.statusCode, 404);
    assert.equal(err.code, "server_error");
  });

  it("PayNetworkError has code", () => {
    const err = new PayNetworkError("connection refused");
    assert.equal(err.code, "network_error");
  });

  it("PayInsufficientFundsError includes fund link hint", () => {
    const err = new PayInsufficientFundsError("low balance", 5, 10);
    assert.ok(err.message.includes("createFundLink"));
    assert.equal(err.balance, 5);
    assert.equal(err.required, 10);
    assert.equal(err.code, "insufficient_funds");
  });
});
